List fix: no reload

andras-tim opened the pull request:
**By submitting this pull request, I confirm the following:** 
*please fill any appropriate checkboxes, e.g: [X]*

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [X] I have made only one major change in my proposed changes.
- [X] I have commented my proposed changes within the code.
- [X] I have tested my proposed changes, and have included unit tests where possible.
- [X] I am willing to help maintain this change if there are issues with it later.
- [X] I give this submission freely and claim no ownership.
- [X] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [X] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

Please make sure you [Sign Off](https://github.com/pi-hole/pi-hole/wiki/How-to-signoff-your-commits.) all commits. Pi-hole enforces the [DCO](https://github.com/pi-hole/pi-hole/wiki/Contributing-to-the-project).

---
**What does this PR aim to accomplish?:**
Fix the `--noreload` flag in the `list.sh`


**How does this PR accomplish the above?:**
* Split the responsibility of the `reload` variable.


**What documentation changes (if any) are needed to support this PR?:**
I don't think this is necessary now.


---
* You must follow the template instructions. Failure to do so will result in your pull request being closed.
* Please respect that Pi-hole is developed by volunteers, who can only reply in their spare time.

PromoFaux commented:
Thanks for the submission!

Is there a specific issue / discourse posting that this PR is aiming to fix? Could you please add a brief explanation as to what happens before this changeset, and what happens after? 

Thanks!

andras-tim commented:
> Thanks for the submission!
> 
> Is there a specific issue / discourse posting that this PR is aiming to fix? Could you please add a brief explanation as to what happens before this changeset, and what happens after?
> 
> Thanks!

Dear @PromoFaux,

Your welcome! :wink: I didn't discuss this issue w/ anyone, yet.

I wrote an Ansible Playbook for the home setup and there I wanted to do proper mass white-/blacklist management. I aimed to avoid the unnecessary reloading when I know I will put more records to the SQLite.

The original behavior:

```
+ pihole -b --nuke
  [i] blacklist already empty. Nothing to do!
+ pihole --wild --nuke
  [✓] Removed 1 domain(s) from the regex blacklist
+ pihole --regex --nuke
  [i] regex blacklist already empty. Nothing to do!
+ pihole -w --nuke
  [✓] Removed 1 domain(s) from the whitelist
+ pihole --white-wild --nuke
  [✓] Removed 2 domain(s) from the regex whitelist
+ pihole --white-regex --nuke
  [i] regex whitelist already empty. Nothing to do!
+ pihole --regex -nr '*;querytype=aaaa' --comment 'Disable IPv6 queries'
  [i] Adding *;querytype=aaaa to the regex blacklist...
  [✓] Reloading DNS lists
+ pihole -w -nr img.alibaba.com
  [i] Adding img.alibaba.com to the whitelist...
  [✓] Reloading DNS lists
+ pihole --white-wild -nr googleadservices.com cib.hu
  [i] Adding (^|\.)googleadservices\.com$ to the regex whitelist...
  [i] Adding (^|\.)cib\.hu$ to the regex whitelist...
  [✓] Reloading DNS lists
+ pihole restartdns reload-lists
  [✓] Reloading DNS lists
```

The new behavior:
``` 
...
+ pihole --white-regex --nuke
  [i] regex whitelist already empty. Nothing to do!
+ pihole --regex -nr '*;querytype=aaaa' --comment 'Disable IPv6 queries'
  [i] Adding *;querytype=aaaa to the regex blacklist...
+ pihole -w -nr img.alibaba.com
  [i] Adding img.alibaba.com to the whitelist...
+ pihole --white-wild -nr googleadservices.com cib.hu
  [i] Adding (^|\.)googleadservices\.com$ to the regex whitelist...
  [i] Adding (^|\.)cib\.hu$ to the regex whitelist...
+ pihole restartdns reload-lists
  [✓] Reloading DNS lists
```
_(no "Reloading DNS lists" when I passed the `-nr` argument)_

Of course, I know, this has not huge performance/time penalty - but I'm a developer also - so I couldn't step over that :slightly_smiling_face:

SlothCroissant commented:
Thanks for this! 

Was going to come leave a new issue for this today - the `-nr` or `--noreload` still tosses the `[✓] Reloading DNS lists` log into the console. Not sure if it's just cosmetic, but I was also looking to script and seeing the extra console lines when they shouldn't be there.

PromoFaux commented on advanced/Scripts/list.sh:
```diff
@@ -258,7 +259,7 @@ while (( "$#" )); do
         "--wild" | "wildcard" ) typeId=3; wildcard=true;;
         "--regex" | "regex"   ) typeId=3;;
-        "-nr"| "--noreload"  ) reload=false;;
+        "-nr"| "--noreload"  ) canReload=false;;
```
```suggestion
        "-nr"| "--noreload"  ) noReloadRequested=true;;
```

PromoFaux commented on advanced/Scripts/list.sh:
```diff
@@ -284,6 +285,6 @@ if $web; then
 fi
 
-if [[ "${reload}" != false ]]; then
+if [[ ${reload} == true ]] && [[ ${canReload} == true ]]; then
```
```suggestion
if [[ ${reload} == true ]] && [[ ${noReloadRequested} == false ]]; then
```

PromoFaux commented on advanced/Scripts/list.sh:
```diff
@@ -13,7 +13,7 @@ basename=pihole
 gravityDBfile="${piholeDir}/gravity.db"
 
-reload=false
+canReload=true
```
```suggestion
noReloadRequested=false
```

PromoFaux reviewed (changes requested):
Sorry for the delay in review - life gets in the way sometimes.

Please look over my suggestions - I think from a logic/code-readability point of view it would be clearer to flip the values and rename the variable to make it explicitly clear what it is doing

andras-tim commented:
> Sorry for the delay in review - life gets in the way sometimes.
> 
> Please look over my suggestions - I think from a logic/code-readability point of view it would be clearer to flip the values and rename the variable to make it explicitly clear what it is doing

I'm sorry I answered for so long. Generally, I prefer to avoid the negated-voice, but I did minimalize the change.

So, thanks for the suggestions.

andras-tim commented:
@PromoFaux, should I merge the fixup patches?

MichaIng commented:
Nice, makes sense. In bash syntax conditionals, such things can be merged btw:
```sh
[[ ${reload} == true ]] && [[ ${noReloadRequested} == false ]]
```
=>
```sh
[[ ${reload} == true && ${noReloadRequested} == false ]]
```
If you have a change to squash the commits for a cleaner history in the main branch, that would be great it think. Otherwise this should be good to go?

dschaper commented:
The development branch has had some changes recently. A rebase of the commits on to the branch head and a force push after a squash of the commits would be nice. 

We can override DCO if the submitter adds a sign off statement in a message here or in the PR message. Otherwise https://github.com/pi-hole/pi-hole/pull/3981/checks?check_run_id=2150632344 has the steps needed to sign and force-push in the signature.

andras-tim commented:
Sorry for the latency. I merged the update patches, fixed the signoff, merged the conditions according to the suggestion of @MichaIng, and rebased the whole stuff to the current master. I think I'm done now :wink:
